REGRESSION (Safari 16): Service worker clientId is empty for web worker fetches

youennf opened the pull request:
#### 6facf7d89d16cabe9fddd73cdfc28cb2881c40de
<pre>
REGRESSION (Safari 16): Service worker clientId is empty for web worker fetches
<a href="https://bugs.webkit.org/show_bug.cgi?id=245425">https://bugs.webkit.org/show_bug.cgi?id=245425</a>
rdar://problem/100368228

Reviewed by Chris Dumez.

After introducing correct matching of service worker registration for worker loads, we correctly filled resultingClientId to the worker context id.
As per <a href="https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-classic-worker-script">https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-classic-worker-script</a>, we need to set clientId to the context Id creating the worker.
To do so, we introduce resultingClientId as a FetchOptions so that we can set both clientId and resultingClientId when doing a load.
Set resultingClientId in DocumentLoader and set both clientId and resultingClientId for workers.
Update networking process code to pass clientId and resultingClientId when starting a fetch event.

Drive=by fix, updateAppInitiatedValue needs to hop to main thread.

Covered by newly added test.

* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/sample-worker-interceptor.js:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/worker-interception.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/worker-interception.https.html:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/service-workers/service-worker/worker-interception.https-expected.txt:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::loadMainResource):
* Source/WebCore/loader/FetchOptions.h:
(WebCore::FetchOptions::encode const):
(WebCore::FetchOptions::decode):
* Source/WebCore/workers/WorkerScriptLoader.cpp:
(WebCore::WorkerScriptLoader::loadAsynchronously):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::startFetch):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::controlClient):
(WebKit::WebSWServerConnection::createFetchTask):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::updateAppInitiatedValue):

Canonical link: <a href="https://commits.webkit.org/254944@main">https://commits.webkit.org/254944@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8b57c4a23f99d4527067d5ed100bf60d8381b68

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90683 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35263 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21281 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99998 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/158320 "Reverted pull request changes (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94692 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33761 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28896 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83041 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96397 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96338 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26875 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77506 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26711 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81651 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81420 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69739 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34853 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15457 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32665 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16437 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3448 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36431 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39375 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38352 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35526 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/FetchOptions.h:
```diff
@@ -62,6 +62,7 @@ struct FetchOptions {
     String integrity;
     bool keepAlive { false };
     std::optional<ScriptExecutionContextIdentifier> clientIdentifier;
+    std::optional<ScriptExecutionContextIdentifier> resultingClientIdentifier;
```
Might be worth a comment to explain the difference with clientIdentifier?

youennf replied:
Let's add something like
```
// id of https://fetch.spec.whatwg.org/#concept-request-client
// id of https://fetch.spec.whatwg.org/#concept-request-reserved-client
```

cdumez reviewed (approved):
LGTM

youennf committed:
REGRESSION (Safari 16): Service worker clientId is empty for web worker fetches
https://bugs.webkit.org/show_bug.cgi?id=245425
rdar://problem/100368228

Reviewed by Chris Dumez.

After introducing correct matching of service worker registration for worker loads, we correctly filled resultingClientId to the worker context id.
As per https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-classic-worker-script, we need to set clientId to the context Id creating the worker.
To do so, we introduce resultingClientId as a FetchOptions so that we can set both clientId and resultingClientId when doing a load.
Set resultingClientId in DocumentLoader and set both clientId and resultingClientId for workers.
Update networking process code to pass clientId and resultingClientId when starting a fetch event.

Drive=by fix, updateAppInitiatedValue needs to hop to main thread.

Covered by newly added test.

* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/sample-worker-interceptor.js:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/worker-interception.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/worker-interception.https.html:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/service-workers/service-worker/worker-interception.https-expected.txt:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::loadMainResource):
* Source/WebCore/loader/FetchOptions.h:
(WebCore::FetchOptions::encode const):
(WebCore::FetchOptions::decode):
* Source/WebCore/workers/WorkerScriptLoader.cpp:
(WebCore::WorkerScriptLoader::loadAsynchronously):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::startFetch):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::controlClient):
(WebKit::WebSWServerConnection::createFetchTask):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::updateAppInitiatedValue):

Canonical link: https://commits.webkit.org/254944@main
